packages/drivers/driver-base/src/documentDeltaConnection.ts

```diff
@@ -368,12 +357,17 @@ export class DocumentDeltaConnection
 
                 this.removeConnectionListeners();
                 resolve(response);
-                success = true;
             });
 
-            // WARNING: this has to stay as addTrackedListener listener and not be removed after successful connection.
-            // Reason: this.on() implementation does not subscribe to "error" socket events to propagate it to consumers
-            // of this class - it relies on this code to do so.
+            // Socket can be disconnected while waiting for Fluid protocol messages
+            // (connect_document_error / connect_document_success), as well as before DeltaManager
+            // had a chance to register its handlers.
+            this.addTrackedListener("disconnect", (reason) => {
+                const err = this.createErrorObject("disconnect", reason);
+                this.emit("disconnect", err);
+                fail(true, err);
```

Comment:
This will cause the connection to dispose of itself (even after successful connection established).  Although this is the state we want to get into ultimately after a disconnect, with this change it becomes more confusing to me who is "in control" of the connection's lifecycle after successful connection.  E.g. `DeltaManager` is still going to try calling `dispose()`  despite it not being necessary.  The `DoubleClose` log you're removing hints that this ambiguity has been a source of bugs in the past -- how can we clarify the lifecycle ownership?

Reply:
DM can dispose connection at any time for any reason, that can't be taken away (we do that in non-error cases).
Connection object has to do proper cleanup if error / disconnect happened before successful handshake. We also need to do some tracking / cleaning while connection object is in transition (i.e. DM has not had a chance to see it).

I'm not sure I see good alternative here.
We can follow patter of early "op" handler (i.e. subscribe and buffer events until "real" handler shows up), but it's a source of its own bugs.

We can revert flow and delay connection until DM had a chance to install all the handlers, but that's quite a big change (need to move around a lot of code, including error handling code in connectCore, OdspDocumentDeltaConnection.create, OdspDocumentService.connectToDeltaStream - I'd rather do it not under pressure of releasing TAP.

From what I can observe (tracing code and looking at stress test results), the issue described above DoubleClose comment is no longer valid as we allow multiple disposition by making non-first disposing noop. The only problem I see in this code (that exists before this PR) is that dispose() does not raise "disconnect" event as we clear all event handlers first. It's not a problem for fail() flow as caller raises proper event before calling fail(), but it is something that is not great about this code and we can improve.

Reply:
BTW, it's not much different RE socket/socket.io itself - socket may lose connection and close itself, or client can close it.